interfaces/prompting/requestrules/requestrules.go

```diff
@@ -0,0 +1,877 @@
+				if rule.Timestamp.After(conflictingRule.Timestamp) {
+					rdb.removeRulePermissionFromTree(conflictingRule, conflictingPermission) // must return nil
+					var data map[string]string
+					if conflictingRule.removePermission(conflictingPermission) == prompting.ErrPermissionsListEmpty {
+						rdb.removeRuleWithID(conflictingID)
+						data = map[string]string{"removed": "conflict"}
+					}
+					modifiedUserRuleIDs[conflictingRule.User][conflictingID] = data
+				} else {
+					rule.removePermission(conflictingPermission) // ignore error
+					var data map[string]string
+					modifiedUserRuleIDs[rule.User][rule.ID] = data
+				}
+				needToSave = true
+			}
+		}
+		if len(rule.Constraints.Permissions) > 0 {
+			rdb.addRule(rule)
+		} else {
+			// TODO: record status of the rule ("removed") in modifiedUserRuleIDs
+			needToSave = true
+			data := map[string]string{"removed": "conflict"}
+			modifiedUserRuleIDs[rule.User][rule.ID] = data
+		}
+	}
+}
+
+// PopulateNewRule creates a new Rule with the given contents.
+//
+// Users of requestrules should probably autofill rules from JSON and never call
+// this function directly.
+//
+// Constructs a new rule with the given parameters as values, with the
+// exception of duration. Uses the given duration, in addition to the current
+// time, to compute the expiration time for the rule, and stores that as part
+// of the rule which is returned. If any of the given parameters are invalid,
+// returns a corresponding error.
+//
+// TODO: unexport (probably, avoid confusion with AddRule)
+func (rdb *RuleDB) PopulateNewRule(user uint32, snap string, iface string, constraints *prompting.Constraints, outcome prompting.OutcomeType, lifespan prompting.LifespanType, duration string) (*Rule, error) {
+	if err := constraints.ValidateForInterface(iface); err != nil {
+		return nil, err
+	}
+	if _, err := outcome.AsBool(); err != nil {
+		// This should not occur, since PopulateNewRule should only be called
+		// on values which were validated while unmarshalling
+		return nil, err
+	}
+	id, _ := rdb.maxIDMmap.NextID()
+	currTime := time.Now()
+	expiration, err := lifespan.ParseDuration(duration, currTime)
+	if err != nil {
+		return nil, err
+	}
+	newRule := Rule{
+		ID:          id,
+		Timestamp:   currTime,
+		User:        user,
+		Snap:        snap,
+		Interface:   iface,
+		Constraints: constraints,
+		Outcome:     outcome,
+		Lifespan:    lifespan,
+		Expiration:  expiration,
+	}
+	return &newRule, nil
+}
+
+// IsPathAllowed checks whether the given path with the given permission is
+// allowed or denied by existing rules for the given user, snap, and interface.
+// If no rule applies, returns ErrNoMatchingRule.
+func (rdb *RuleDB) IsPathAllowed(user uint32, snap string, iface string, path string, permission string) (bool, error) {
+	rdb.mutex.Lock()
+	defer rdb.mutex.Unlock()
+	needToSave := false
+	defer func() {
+		if needToSave {
+			rdb.save()
```

Comment:
again we're theoretically only doing a query which is only reading from the DB, but in fact we also try to patch things up as we go

Reply:
Should be fixed now